src/codegen/enums.rs

```diff
@@ -168,9 +168,10 @@ fn generate_enum(
         version_condition(w, env, enum_.version, false, 0)?;
         writeln!(
             w,
-            "impl fmt::Display for {0} {{\n\
-             \tfn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {{\n\
-             \t\twrite!(f, \"{0}::{{}}\", match *self {{",
+            "impl fmt::Display for {0} {{
+\tfn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {{
+\t\twrite!(f, \"{0}::{{}}\", match *self {{
+",
```

Comment:
Why? It looks nicer to have it aligned properly and the output is the same

Reply:
I don't find `\n\` particularly readable, but then again that allows to have leading spaces as indentation.

Would this work? (Notice the `"\`):

```rust
            "\
impl fmt::Display for {0} {{
\tfn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {{
\t\twrite!(f, \"{0}::{{}}\", match *self {{
",
```

That also allows `\t` to be substituted with `    `.

Reply:
That would also work for me (also with four leading spaces).

Just it would be nice to have the code in the string consistently indented, otherwise it's confusing to read.

Reply:
> also with four leading spaces

That would resurface the need for trailing `\n\` again. I'll remove these changed from the PR as they're merely extra noise.

Reply:
Ok :) Maybe let's handle that separately then, @GuillaumeGomez probably also has opinions on this topic

Reply:
It appears both ways are used, one with `\n\` and proper indentation to follow the rest of the code, others with mismatched indentation on the first line and the rest justified all the way to the left. Kept the changes that turn:
```rust
    writeln!(
        w,
        "{}",
        "    #[doc(hidden)]
    __Unknown(i32),
}
"
```
into:
```rust
    writeln!(
        w,
        "\
    #[doc(hidden)]
    __Unknown(i32),
}}
"
```
As per your alignment request.

Reply:
I don't mind to have `\n`. Please keep the indent using whitespaces. :)

Reply:
I don't know how much we should care about formatting (at least indentation) of generated output (seeing a PR to switch to `quote` earlier this year being closed) as the resulting files are pulled through rustfmt anyway.

Reply:
Not much, what matters is that it's readable inside the `gir` code. The autogenerated code itself will be cleaned up anyway, as you say.

Reply:
It seems that'd improve a lot with `quote!` and the way it can expand arrays or iterators, and in general colorize target code in the IDE. But it appears the need for `//` comments with nested `/**/` completely breaks this unfortunately :(.

Reply:
No, `quote!` didn't improve the readability at all.

Reply:
Yeah I tried to use quote here but it only made things worse in practice. Maybe I just used it wrongly though...

Reply:
I'd argue:

```rust
        let type_name = format_ident!("{}", enum_.name);

        let members = members.iter().map(|m| {
            let name = &m.name;
            let ident = format_ident!("{}", name);
            // TODO: version_condition_tokens(env, m.version, false, 3) -> Option<TokenStream>
            quote!(Self::#ident => #name)
        });

        quote! {
            impl fmt::Display for #type_name {
                fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
                    match *self {
                        #( #members, )*
                        _ => "Unknown",
                    }
                }
            }
        }
```
Is way easier to read than:
```rust
        writeln!(
            w,
            "impl fmt::Display for {0} {{\n\
             \tfn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {{\n\
             \t\twrite!(f, \"{0}::{{}}\", match *self {{",
            enum_.name
        )?;
        for member in &members {
            version_condition(w, env, member.version, false, 3)?;
            writeln!(w, "\t\t\t{0}::{1} => \"{1}\",", enum_.name, member.name)?;
        }
        writeln!(
            w,
            "\t\t\t_ => \"Unknown\",\n\
             \t\t}})\n\
             \t}}\n\
             }}\n"
        )?;
```
But then there are quite a few things that are not ergonomic:

- Not able to expand members in-line (ie. `#(m.name)`);
- Not able to convert strings to identifiers within the macro.

These in turn make it impossible to have something like (pseudo-syntax) `#( Self::#(/*turn into identifier*/ #(members).name) => #(#(members).name), )*` directly in place of the `map`.

Other than that the ability to expand iterators and deal with `Option<>` in expansion is pretty neat.